### ISSUE TITLE
e2guardian.service: Main process exited, code=killed, status=6/ABRT  32-bit system

### ISSUE THREAD

**Issue by KryptonTux** [e2guardian/e2guardian#561]
Hi, i have a problem with a e2guardian in debian Buster. When i boot the service it shows the following. 

● e2guardian.service - E2guardian Web filtering
   Loaded: loaded (/lib/systemd/system/e2guardian.service; enabled; vendor preset: enabled)
   Active: failed (Result: signal) since Tue 2019-11-19 18:39:06 CET; 1s ago
     Docs: http://e2guardian.org/
  Process: 6038 ExecStart=/usr/sbin/e2guardian (code=exited, status=0/SUCCESS)
 Main PID: 6039 (code=killed, signal=ABRT)

nov 19 18:39:05 f0 systemd[1]: Starting E2guardian Web filtering...
nov 19 18:39:06 f0 systemd[1]: Started E2guardian Web filtering.
nov 19 18:39:06 f0 systemd[1]: e2guardian.service: Main process exited, code=killed, status=6/ABRT
nov 19 18:39:06 f0 systemd[1]: e2guardian.service: Failed with result 'signal'.

 dpkg -l | grep -v ii
Deseado=desconocido(U)/Instalar/eliminaR/Purgar/retener(H)
| Estado=No/Inst/ficheros-Conf/desempaqUetado/medio-conF/medio-inst(H)/espera-disparo(W)/pendienTe-disparo
|/ Err?=(ninguno)/requiere-Reinst (Estado,Err: mayúsc.=malo)
||/ Nombre                                    Versión                             Arquitectura Descripción
+++-=========================================-===================================-============-===============================================================================

The package is well installed.
dpkg -l | grep -i e2guardian
ii  e2guardian                                5.3.1-1                             i386  

I have a giga of RAM
              total        used        free      shared  buff/cache   available
Mem:            927         495          96          11         335         284
Swap:          1906         246        1660

I will add all the information you request
Has anyone had a similar problem with e2guardian?
Thanks

**Comment by schallee** [comment]
I believe this is a duplicate of #547. See the possible work around there.

**Comment by fredbcode** [comment]
e2guardian -N please

My package are not compatible with buster, I guess some libraries are missing

**Comment by KryptonTux** [comment]
Schallee The #547 solution is not valid in this case, it has already been tried.

Right after restarting the service

● e2guardian.service - E2guardian Web filtering
   Loaded: loaded (/lib/systemd/system/e2guardian.service; enabled; vendor preset: enabled)
   Active: failed (Result: signal) since Thu 2019-11-28 11:28:21 CET; 1min 54s ago
     Docs: http://e2guardian.org/
  Process: 18005 ExecStart=/usr/sbin/e2guardian (code=exited, status=0/SUCCESS)
 Main PID: 18006 (code=killed, signal=ABRT)

nov 28 11:28:21 f0 systemd[1]: Starting E2guardian Web filtering...
nov 28 11:28:21 f0 systemd[1]: Started E2guardian Web filtering.
nov 28 11:28:21 f0 systemd[1]: e2guardian.service: Main process exited, code=killed, status=6/
nov 28 11:28:21 f0 systemd[1]: e2guardian.service: Failed with result 'signal'.

attached
:~# e2guardian -N
terminate called after throwing an instance of 'std::system_error'
  what():  Resource temporarily unavailable
Aborted



:~# apt-cache depends e2guardian
e2guardian
  PreDepende: init-system-helpers
  Depende: adduser
  Depende: clamav-daemon
  Depende: perl
  Depende: libc6
  Depende: libgcc1
  Depende: libpcre3
  Depende: libssl1.1
  Depende: libstdc++6
  Depende: zlib1g
  Sugiere: squid

The libraries are already proven, and I think they are fine. Do you see any that is not right?

Thanks

**Comment by fredbcode** [comment]
Yes libevent-pthreads-2.1-6

-----

**Comment by fredbcode** [comment]
Wait, which package are you using ?
-- 
Envoyé de mon appareil Android avec Courriel K-9 Mail. Veuillez excuser ma brièveté.

**Comment by KryptonTux** [comment]
I am using version 5.3.1-1
:~# dpkg -l | grep -i e2guardian
ii  e2guardian                                5.3.1-1                             i386         Web content filtering (Dansguardian fork)

**Comment by KryptonTux** [comment]
attached libraries

:~# ldd /usr/sbin/e2guardian
	linux-gate.so.1 (0xb7f67000)
	libpcreposix.so.3 => /usr/lib/i386-linux-gnu/libpcreposix.so.3 (0xb7e36000)
	libpcre.so.3 => /lib/i386-linux-gnu/libpcre.so.3 (0xb7dbd000)
	libpthread.so.0 => /lib/i386-linux-gnu/libpthread.so.0 (0xb7d9c000)
	libz.so.1 => /lib/i386-linux-gnu/libz.so.1 (0xb7d7d000)
	libssl.so.1.1 => /usr/lib/i386-linux-gnu/libssl.so.1.1 (0xb7ce6000)
	libcrypto.so.1.1 => /usr/lib/i386-linux-gnu/libcrypto.so.1.1 (0xb7a26000)
	libstdc++.so.6 => /usr/lib/i386-linux-gnu/libstdc++.so.6 (0xb78a7000)
	libm.so.6 => /lib/i386-linux-gnu/libm.so.6 (0xb77a1000)
	libgcc_s.so.1 => /lib/i386-linux-gnu/libgcc_s.so.1 (0xb7783000)
	libc.so.6 => /lib/i386-linux-gnu/libc.so.6 (0xb75a5000)
	/lib/ld-linux.so.2 (0xb7f69000)
	libdl.so.2 => /lib/i386-linux-gnu/libdl.so.2 (0xb759d000)

**Comment by KryptonTux** [comment]
libevent-pthreads-2.1-6 It is already in its most recent version (2.1.8-stable-4)

**Comment by fredbcode** [comment]
You mean from debian, official package ?

**Comment by fredbcode** [comment]
How many workers ?

**Comment by KryptonTux** [comment]
> You mean from debian, official package ?

Yes. I use official package.


> How many workers ?

:~# systemctl start e2guardian
top -  1 user,  load average: 0,08, 0,10, 0,03
Tasks: 268 total,   1 running, 267 sleeping,   0 stopped,   0 zombie
%Cpu(s):  0,5 us,  1,2 sy,  0,0 ni, 98,3 id,  0,0 wa,  0,0 hi,  0,0 si,  0,0 st
MiB Mem :    927,6 total,    228,7 free,    184,1 used,    514,7 buff/cache
MiB Swap:   1907,0 total,   1350,0 free,    557,0 used.    579,3 avail Mem 

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND                                                                           
  649 root        20   0      143604  856   380 S   1,7   0,1  77:20.37 collectd                                                                          
17757 root      20   0       0       0      0 I   0,7      0,0   0:06.25   kworker/1:2-events                                                                
19815 root      20   0       0       0      0 I   0,7      0,0   0:01.64   kworker/0:2-events                                                                
20418 root      20   0       10884    3668   3044 R   0,3   0,4      0:00.14 top                                                                               
21365 root      20   0       216520  26960  20156 S   0,3   2,8   0:25.82 apache2                                                                           
27477 proxy    20   0       64760   8044   2948 S   0,3   0,8      1:08.94 squid                                                                             
         1 root      20   0      36712   5536   3916 S   0,0   0,6       2:02.81 systemd

**Comment by fredbcode** [comment]
I mean workers value in e2guardian.conf 

If there is no change in your configuration, I'm afraid that the Debian package is just buggy

I will made an official package for the latest Debian, soon I hope ...

-- 
Envoyé de mon appareil Android avec Courriel K-9 Mail. Veuillez excuser ma brièveté.

**Comment by KryptonTux** [comment]
Attached workers e2guardian.conf

:~# grep -v "#" /etc/e2guardian/e2guardian.conf | sed -e '/^$/d'
reportinglevel = 3
languagedir = '/usr/share/e2guardian/languages'
language = 'spanish'
loglevel = 2 
logexceptionhits = 0
logfileformat = 1 
filterip =
filterports = 8080
transparenthttpsport = 8443
proxyip = 127.0.0.1
proxyport = 3128
proxytimeout = 20
connecttimeout = 3
proxyexchange = 20
pcontimeout = 55
usecustombannedimage = on
custombannedimagefile = '/usr/share/e2guardian/transparent1x1.gif'
usecustombannedflash = on
custombannedflashfile = '/usr/share/e2guardian/blockedflash.swf'
filtergroups = 1
filtergroupslist = '/etc/e2guardian/lists/filtergroupslist'
iplist = 'name=bannedclient,messageno=100,logmessageno=103,path=/etc/e2guardian/lists/bannediplist'
iplist = 'name=exceptionclient,messageno=600,path=/etc/e2guardian/lists/exceptioniplist'
iplist = 'name=authexception,messageno=602,path=/etc/e2guardian/lists/authexceptioniplist'
sitelist = 'name=authexception,messageno=602,path=/etc/e2guardian/lists/authexceptionsitelist'
urllist = 'name=authexception,messageno=603,path=/etc/e2guardian/lists/authexceptionurllist'
showweightedfound = off
weightedphrasemode = 2
phrasefiltermode = 2
preservecase = 0
hexdecodecontent = off
forcequicksearch = off
reverseaddresslookups = off
reverseclientiplookups = off
logclienthostnames = off
maxcontentfiltersize = 256
maxcontentramcachescansize = 2000
maxcontentfilecachescansize = 20000
filecachedir = '/tmp'
deletedownloadedtempfiles = on
initialtrickledelay = 20
trickledelay = 10
downloadmanager = '/etc/e2guardian/downloadmanagers/default.conf'
contentscannertimeout = 60
contentscanexceptions = off
recheckreplacedurls = off
forwardedfor = off
usexforwardedfor = off
logconnectionhandlingerrors = on
httpworkers = 500 
nodaemon = off
nologger = off
logadblocks = off
loguseragent = off
enablessl = off
preauthstoryboard = '/etc/e2guardian/preauth.story'
searchsitelistforip = on

Thanks for your time

**Comment by fredbcode** [comment]
httpworkers = 500 is a good value, wrong direction

**Comment by KryptonTux** [comment]
What direction do you mean?

**Comment by KryptonTux** [comment]
httpworkers = 370 and it works fine, but does not accept any higher value. 

The problem occurs with 32-bit servers. But I have others with 64 bits where there is no problem.

The value 500 works on 64-bit servers but not on 32-bit servers

**Comment by fredbcode** [comment]
Please check your systemd script

https://github.com/fredbcode/scripts/blob/master/debian_package/e2debian_package/data/lib/systemd/system/e2guardian.service

I guess some values are missing in debian package

**Comment by philipianpearce** [comment]
Ah!  

I have not worked on 32-bit systems for many years, but this is the likely reason for this:-

By default linux allocates 10M VM for a stack on each tread so 500 x 10m = 5G = too big for 32-bit.

Check the stack size with   ulimit -s

You may be able to get more threads (if you really need them) by reducing the stack size to say 2m with ulimit -s 2048 before starting e2g.   It may be a good idea to do this anyway as it would free up VM for lists and working memory.   (370 x 10 = 3.7G so only leaves 300M as working area for all e2g threads).

If stack VM size is kept at the default then I suggest that you have a max of 300 httpworkers on 32-bit so as to allow enough memory for lists and working memory.

**Comment by KryptonTux** [comment]
> Please check your systemd script https://github.com/fredbcode/scripts/blob/master/debian_package/e2debian_package/data/lib/systemd/system/e2guardian.service I guess some values are missing in debian package

:~# cat /usr/lib/systemd/system/e2guardian.service
[Unit]
Description=E2guardian Web filtering
After=network.target
Documentation=http://e2guardian.org/

[Service]
Type=forking
ExecStart=/usr/sbin/e2guardian

[Install]
WantedBy=multi-user.target



> Ah!
> 
> I have not worked on 32-bit systems for many years, but this is the likely reason for this:-
> 
> By default linux allocates 10M VM for a stack on each tread so 500 x 10m = 5G = too big for 32-bit.
> 
> Check the stack size with ulimit -s
> 
> You may be able to get more threads (if you really need them) by reducing the stack size to say 2m with ulimit -s 2048 before starting e2g. It may be a good idea to do this anyway as it would free up VM for lists and working memory. (370 x 10 = 3.7G so only leaves 300M as working area for all e2g threads).
> 
> If stack VM size is kept at the default then I suggest that you have a max of 300 httpworkers on 32-bit so as to allow enough memory for lists and working memory.


This is very interesting. I can't replace 32-bit servers, I'm working on it.

In addition to the solution of reducing the httpworker to less than 370 I have found another one. But I don't know if it's the most optimal.

Add the LimitSTACK = infinity field to the e2guardian.service file

Can the system have any danger?

Thanks for your time

**Comment by philipianpearce** [comment]
LimitSTACK = infinity will actually set the soft limit to the hard limit.
So use  ulimit -Hs
to see what the hard limit is (in kB).

Otherwise best set LimitSTACK = 2000000 and if you get stack overflow increase it.

**Comment by KryptonTux** [comment]
Finally, the problem has been solved by reducing the httpworkers to 370.

Thank you all for your time.
